src/migrations/00054-migrate-release-versioning-fields.sql

```diff
@@ -0,0 +1,96 @@
+UPDATE "release"
+SET "revision" = NULL
+WHERE "release type" = 'draft'
+	AND "revision" IS NOT NULL;
+
+UPDATE "release"
+SET "is finalized at-date" = "release"."created at"
+WHERE "release type" = 'final'
+	AND "is finalized at-date" IS NULL;
+
+-- Parse the release version (best effort) and split it into semver parts with an increasing revision for identical semvers.
+-- The _logstream variants & 4digit versions will get a higher revision b/c they where created later than the base version.
+-- For release versions that are empty or can't be parsed or were parsed as 0.0.0, set 0.0.0 as the semver and increase the revision.
+WITH "parsed version release" AS (
+	SELECT r0."id",
+		r0."is finalized at-date",
+		r0."belongs to-application",
+		COALESCE(CAST(SUBSTRING((STRING_TO_ARRAY(r0."release version", '.'))[1] FROM '[0-9]+') AS INTEGER), 0) AS "semver major",
+		COALESCE(CAST(SUBSTRING((STRING_TO_ARRAY(r0."release version", '.'))[2] FROM '[0-9]+') AS INTEGER), 0) AS "semver minor",
+		COALESCE(CAST(SUBSTRING((STRING_TO_ARRAY(r0."release version", '.'))[3] FROM '[0-9]+') AS INTEGER), 0) AS "semver patch"
+	FROM "release" r0
+	WHERE r0."release type" = 'final'
+		AND r0."release version" SIMILAR TO 'v?[0-9]+(\.[0-9]+(\.[0-9]+(\.[0-9]+)?(_logstream[0-9]?)?)?)?'
+),
+"revisioned release" AS (
+	SELECT splitedReleaseVersion."id",
+		splitedReleaseVersion."semver major",
+		splitedReleaseVersion."semver minor",
+		splitedReleaseVersion."semver patch",
+		(ROW_NUMBER() OVER (
+			PARTITION BY splitedReleaseVersion."belongs to-application",
+				splitedReleaseVersion."semver major",
+				splitedReleaseVersion."semver minor",
+				splitedReleaseVersion."semver patch"
+			ORDER BY
+				splitedReleaseVersion."is finalized at-date" ASC,
+				splitedReleaseVersion."id" ASC
+		) - 1) AS "revision"
+	FROM (
+		SELECT pvr."id",
+			pvr."is finalized at-date",
+			pvr."belongs to-application",
+			pvr."semver major",
+			pvr."semver minor",
+			pvr."semver patch"
+		FROM "parsed version release" pvr
+		UNION ALL
+		SELECT unparsedRelease."id",
+			unparsedRelease."is finalized at-date",
+			unparsedRelease."belongs to-application",
+			0 AS "semver major",
+			0 AS "semver minor",
+			0 AS "semver patch"
+		FROM "release" unparsedRelease
+		WHERE unparsedRelease."release type" = 'final'
+			AND NOT EXISTS (
+				SELECT 1
+				FROM "parsed version release" pvr
+				WHERE pvr."id" = unparsedRelease."id"
+			)
+	) AS splitedReleaseVersion
+)
+UPDATE "release" r  SET
```

Comment:
FYI that the SELECT of this takes about 4.5 seconds on chartio, updating almost all 1,7m releases.
> Hash Join (cost=3334918.49..3428992.14 rows=1710430 width=61) (actual time=3078.994..4267.474 rows=1007355 loops=1)

Reply:
Hey @thgreasi it's not immediately clear to me where this `Hash Join` happens. Can you please post the full query plan? 
(no promise if I would have a better idea though :sweat_smile:)

On a related note: What's the expected performance here? 
Sounds like 4.5 seconds is too long. But I imagine that it might go faster on the production db because of different setup.

Reply:
@gelbal that was from a SELECT query I was doing on chartio, it should be from the join between the `r` and the `rr`.
Here is the full plan for the UPDATE as I got it from my fasttest DB:
```
Update on release r  (cost=20.77..31.07 rows=2 width=2810) (actual time=0.003..0.004 rows=0 loops=1)
  CTE parsed version release
    ->  Seq Scan on release r0  (cost=0.00..10.33 rows=1 width=28) (never executed)
          Filter: ((("release version")::text ~ '^(?:v?[0-9]+(?:\.[0-9]+(?:\.[0-9]+(?:\.[0-9]+)?(?:.logstream[0-9]?)?)?)?)$'::text) AND (("release type")::text = 'final'::text))
  ->  Hash Join  (cost=10.44..20.74 rows=2 width=2810) (actual time=0.002..0.003 rows=0 loops=1)
        Hash Cond: (r.id = rr.id)
        ->  Seq Scan on release r  (cost=0.00..10.20 rows=20 width=2746) (actual time=0.002..0.002 rows=0 loops=1)
        ->  Hash  (cost=10.42..10.42 rows=2 width=72) (never executed)
              ->  Subquery Scan on rr  (cost=10.33..10.42 rows=2 width=72) (never executed)
                    ->  WindowAgg  (cost=10.33..10.40 rows=2 width=36) (never executed)
                          ->  Sort  (cost=10.33..10.34 rows=2 width=28) (never executed)
                                Sort Key: pvr."belongs to-application", pvr."semver major", pvr."semver minor", pvr."semver patch", pvr."is finalized at-date", pvr.id
                                ->  Append  (cost=0.00..10.32 rows=2 width=28) (never executed)
                                      ->  CTE Scan on "parsed version release" pvr  (cost=0.00..0.02 rows=1 width=28) (never executed)
                                      ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..10.29 rows=1 width=28) (never executed)
                                            ->  Nested Loop Anti Join  (cost=0.00..10.28 rows=1 width=28) (never executed)
                                                  Join Filter: (pvr_1.id = unparsedrelease.id)
                                                  ->  Seq Scan on release unparsedrelease  (cost=0.00..10.25 rows=1 width=16) (never executed)
                                                        Filter: (("release type")::text = 'final'::text)
                                                  ->  CTE Scan on "parsed version release" pvr_1  (cost=0.00..0.02 rows=1 width=4) (never executed)
Planning Time: 1.921 ms
Execution Time: 0.285 ms
```

and here is the respective for a query that just changes the UPDATE with a SELECT from chartio
```sql
-- everything up to here was the same, except that I removed "is finalized at-date" b/c it isn't in the external schema
SELECT *
FROM "release" r,
	"revisioned release" rr
WHERE r."id" = rr."id";
```
```
QUERY PLAN
Merge Join  (cost=2646613.98..3817498.92 rows=1835109 width=1538) (actual time=3254.764..5057.455 rows=1007716 loops=1)
  Merge Cond: (release.id = rr.id)
  CTE parsed version release
    ->  Gather  (cost=1000.00..1059884.90 rows=16430 width=20) (actual time=179.211..291.430 rows=10320 loops=1)
          Workers Planned: 4
          Workers Launched: 4
          ->  Parallel Seq Scan on release release_2  (cost=0.00..1057241.90 rows=4108 width=20) (actual time=176.554..291.249 rows=2064 loops=5)
                Filter: ((("release version")::text ~ '^(?:v?[0-9]+(?:\.[0-9]+(?:\.[0-9]+(?:\.[0-9]+)?(?:.logstream[0-9]?)?)?)?)$'::text) AND (("release type")::text = 'final'::text))
                Rows Removed by Filter: 199479
  ->  Index Scan using release_pkey on release  (cost=0.43..1092893.24 rows=1835109 width=1450) (actual time=0.017..970.756 rows=1007716 loops=1)
  ->  Sort  (cost=1586728.66..1591316.43 rows=1835109 width=24) (actual time=3254.736..3395.321 rows=1007716 loops=1)
        Sort Key: rr.id
        Sort Method: quicksort  Memory: 103304kB
        ->  Subquery Scan on rr  (cost=1322404.75..1395809.11 rows=1835109 width=24) (actual time=2262.859..2994.351 rows=1007716 loops=1)
              ->  WindowAgg  (cost=1322404.75..1377458.02 rows=1835109 width=28) (actual time=2262.858..2885.288 rows=1007716 loops=1)
                    ->  Sort  (cost=1322404.75..1326992.52 rows=1835109 width=20) (actual time=2262.845..2411.443 rows=1007716 loops=1)
                          Sort Key: pvr."belongs to-application", pvr."semver major", pvr."semver minor", pvr."semver patch", pvr.id
                          Sort Method: quicksort  Memory: 103304kB
                          ->  Append  (cost=0.00..1131485.20 rows=1835109 width=20) (actual time=179.217..1555.836 rows=1007716 loops=1)
                                ->  CTE Scan on "parsed version release" pvr  (cost=0.00..328.60 rows=16430 width=20) (actual time=179.216..293.945 rows=10320 loops=1)
                                ->  Subquery Scan on "*SELECT* 2"  (cost=1479.46..1121981.05 rows=1818679 width=20) (actual time=3.368..1181.615 rows=997396 loops=1)
                                      ->  Merge Anti Join  (cost=1479.46..1103794.26 rows=1818679 width=20) (actual time=3.367..1075.599 rows=997396 loops=1)
                                            Merge Cond: (release_1.id = pvr_1.id)
                                            ->  Index Scan using release_pkey on release release_1  (cost=0.43..1097481.01 rows=1835109 width=8) (actual time=0.020..921.823 rows=1007716 loops=1)
                                                  Filter: (("release type")::text = 'final'::text)
                                            ->  Sort  (cost=1479.03..1520.11 rows=16430 width=4) (actual time=3.343..4.986 rows=10320 loops=1)
                                                  Sort Key: pvr_1.id
                                                  Sort Method: quicksort  Memory: 868kB
                                                  ->  CTE Scan on "parsed version release" pvr_1  (cost=0.00..328.60 rows=16430 width=4) (actual time=0.001..1.196 rows=10320 loops=1)
Planning Time: 1.339 ms
Execution Time: 5117.694 ms
```

Reply:
@gelbal  After some experiments with Page, we dropped down to 2.3 seconds!!!
```
SET JIT = 0; EXPLAIN ANALYSE
WITH "parsed version release" AS (
	SELECT r0."id",
		r0."belongs to-application", 
		CASE
			WHEN r0."release version" SIMILAR TO 'v?[0-9]+(.[0-9]+(.[0-9]+(.[0-9]+)?(_logstream[0-9]?)?)?)?' THEN
				COALESCE(CAST(SUBSTRING((STRING_TO_ARRAY(r0."release version", '.'))[1] FROM '[0-9]+') AS INTEGER), 0)
			ELSE 0
		END AS "new semver major",
		CASE
			WHEN r0."release version" SIMILAR TO 'v?[0-9]+(.[0-9]+(.[0-9]+(.[0-9]+)?(_logstream[0-9]?)?)?)?' THEN
				COALESCE(CAST(SUBSTRING((STRING_TO_ARRAY(r0."release version", '.'))[2] FROM '[0-9]+') AS INTEGER), 0)
			ELSE 0
		END AS "new semver minor",
		CASE
			WHEN r0."release version" SIMILAR TO 'v?[0-9]+(.[0-9]+(.[0-9]+(.[0-9]+)?(_logstream[0-9]?)?)?)?' THEN
				COALESCE(CAST(SUBSTRING((STRING_TO_ARRAY(r0."release version", '.'))[3] FROM '[0-9]+') AS INTEGER), 0)
			ELSE 0
		END AS "new semver patch" 
	FROM "release" r0
	WHERE r0."release type" = 'final'
	ORDER BY "belongs to-application",
		"new semver major",
		"new semver minor",
		"new semver patch",
		"id"
),
"revisioned release" AS (
	SELECT split."id",
		split."new semver major" AS "semver major",
		split."new semver minor" AS "semver minor",  
		split."new semver patch" AS "semver patch",
		(ROW_NUMBER() OVER (
			PARTITION BY split."belongs to-application", 
				split."new semver major",
				split."new semver minor",
				split."new semver patch"
			ORDER BY
				split."id" ASC
		) - 1) AS "revision"
	FROM "parsed version release" AS split
)
SELECT r."belongs to-application", rr.*
FROM "release" r,
	"revisioned release" rr
WHERE r."id" = rr."id"
```
```
QUERY PLAN
Hash Join  (cost=1196744.97..1508455.77 rows=1835108 width=28) (actual time=859.507..2243.405 rows=1007837 loops=1)
  Hash Cond: (split.id = release.id)
  ->  WindowAgg  (cost=1117318.00..1405860.55 rows=1835108 width=28) (actual time=457.167..1405.150 rows=1007837 loops=1)
        ->  Subquery Scan on split  (cost=1117318.00..1355395.08 rows=1835108 width=20) (actual time=457.152..921.196 rows=1007837 loops=1)
              ->  Gather Merge  (cost=1117318.00..1337044.00 rows=1835108 width=20) (actual time=457.149..809.078 rows=1007837 loops=1)
                    Workers Planned: 4
                    Workers Launched: 4
                    ->  Sort  (cost=1116317.94..1117464.89 rows=458777 width=20) (actual time=453.423..487.303 rows=201567 loops=5)
"                          Sort Key: release_1.""belongs to-application"", (CASE WHEN ((release_1.""release version"")::text ~ '^(?:v?[0-9]+(?:\.[0-9]+(?:\.[0-9]+(?:\.[0-9]+)?(?:.logstream[0-9]?)?)?)?)$'::text) THEN COALESCE((""substring""((string_to_array((release_1.""release version"")::text, '.'::text))[1], '[0-9]+'::text))::integer, 0) ELSE 0 END), (CASE WHEN ((release_1.""release version"")::text ~ '^(?:v?[0-9]+(?:\.[0-9]+(?:\.[0-9]+(?:\.[0-9]+)?(?:.logstream[0-9]?)?)?)?)$'::text) THEN COALESCE((""substring""((string_to_array((release_1.""release version"")::text, '.'::text))[2], '[0-9]+'::text))::integer, 0) ELSE 0 END), (CASE WHEN ((release_1.""release version"")::text ~ '^(?:v?[0-9]+(?:\.[0-9]+(?:\.[0-9]+(?:\.[0-9]+)?(?:.logstream[0-9]?)?)?)?)$'::text) THEN COALESCE((""substring""((string_to_array((release_1.""release version"")::text, '.'::text))[3], '[0-9]+'::text))::integer, 0) ELSE 0 END), release_1.id"
                          Sort Method: quicksort  Memory: 21911kB
                          Worker 0:  Sort Method: quicksort  Memory: 21996kB
                          Worker 1:  Sort Method: quicksort  Memory: 21962kB
                          Worker 2:  Sort Method: quicksort  Memory: 21795kB
                          Worker 3:  Sort Method: quicksort  Memory: 21796kB
                          ->  Parallel Seq Scan on release release_1  (cost=0.00..1073175.85 rows=458777 width=20) (actual time=0.008..336.905 rows=201567 loops=5)
"                                Filter: ((""release type"")::text = 'final'::text)"
  ->  Hash  (cost=56488.11..56488.11 rows=1835109 width=8) (actual time=398.081..398.081 rows=1007837 loops=1)
        Buckets: 2097152  Batches: 1  Memory Usage: 55753kB
        ->  Index Only Scan using release_id_belongs_to_app_idx on release  (cost=0.43..56488.11 rows=1835109 width=8) (actual time=0.014..168.863 rows=1007837 loops=1)
              Heap Fetches: 8007
Planning Time: 0.802 ms
Execution Time: 2292.437 ms
```

Reply:
Good to hear @thgreasi. I couldn't find the time to take a look before now. Interesting to see that it's faster to repeat that `SIMILAR TO` statements instead of putting it under `WHERE`.